### DISCUSSION TITLE
Python Support (pyallel-consumer?)

### DISCUSSION THREAD

**Discussion by acooke84** [confluentinc/parallel-consumer#820] [Q&A] [answered]
Hello!

It looks as though there was an initiative to add python support for kafka parallel consumers, as shown in the following PR and PyPi package. 

https://github.com/confluentinc/parallel-consumer/pull/443
https://pypi.org/project/pyallel-consumer/

I was wondering if this would ever be revisited or if there are any other known approaches for writing safe parallel consumers in python. 

The only other approach I've read about is having a python API called from java code using the parallel-consumers library, or using some other sort of hybrid approach with something like Jython or JPype or the JNI. (Addressed in this closed discussion a little bit: https://github.com/confluentinc/parallel-consumer/discussions/754)

**Comment by rkolesnev** [comment, marked as answer]
There are no plans to revisit Python integration at this time.
I don't personally think we would block development in that direction if there is community submissions - but it is not something that we plan to work on in foreseeable future.

**Comment by astubbs** [comment]
@acooke84 - it is being revisited, on the maintained fork, and in a different shape from `pyallel-consumer`.

That PR (#443) reimplemented the engine in Python; its branch is preserved in the fork's archive, and the approach was not continued. What the fork is building instead keeps one engine and gives Python two ways to reach it: a sidecar your Python process talks to over a frozen gRPC contract, and an in-process path where the engine runs inside the Python interpreter as a GraalVM shared library, with no gRPC and no JVM install. Both are on open pull requests, not in a release - the wire contract (astubbs/parallel-consumer#383), the sidecar process (astubbs/parallel-consumer#384), the conformance suite every client must pass (astubbs/parallel-consumer#387), the per-language clients including Python (astubbs/parallel-consumer#390), and the in-process Go, Python and C path (astubbs/parallel-consumer#340). The tracking issue is astubbs/parallel-consumer#242, this tracker's #154; the "What comes next" section of [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0) lists it among the open pull requests that follow the release.

Until then, the approaches you listed are the honest state of the art - a Java service in front of a Python one, or a bridge like PemJa - and a comment on astubbs/parallel-consumer#242 saying Python is your case is what decides which client gets finished first.

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
